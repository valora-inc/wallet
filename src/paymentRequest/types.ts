import { NotificationTypes } from 'src/notifications/types'

export enum PaymentRequestStatus {
  REQUESTED = 'REQUESTED',
  COMPLETED = 'COMPLETED',
  DECLINED = 'DECLINED',
  CANCELLED = 'CANCELLED',
}

export interface PaymentRequest {
  uid?: string
  amount: string
  comment?: string
  createdAt: number
  requesterAddress: string
  requesteeAddress: string
  requesterE164Number?: string
  status: PaymentRequestStatus
  notified: boolean
  type?: NotificationTypes.PAYMENT_REQUESTED
}

export interface WriteablePaymentRequest extends Omit<PaymentRequest, 'createdAt'> {
  createdAt: object
}
