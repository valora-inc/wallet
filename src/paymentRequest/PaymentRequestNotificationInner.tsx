import { TFunction } from 'i18next'
import * as React from 'react'
import { getDisplayName, Recipient } from 'src/recipients/recipient'

interface Props {
  amount: string
  recipient: Recipient
  t: TFunction
}

export default function PaymentRequestNotificationInner(props: Props) {
  const { recipient, t } = props
  const displayName = getDisplayName(recipient, t)

  // Using a fragment to suppress a limitation with TypeScript and functional
  // components returning a string
  // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20544
  return <>{displayName}</>
}
