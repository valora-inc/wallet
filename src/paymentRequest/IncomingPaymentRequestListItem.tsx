import { useFocusEffect } from '@react-navigation/native'
import BigNumber from 'bignumber.js'
import React, { useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ActivityIndicator, StyleSheet, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { HomeEvents } from 'src/analytics/Events'
import { SendOrigin } from 'src/analytics/types'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { TokenTransactionType } from 'src/apollo/types'
import ContactCircle from 'src/components/ContactCircle'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import RequestMessagingCard from 'src/components/RequestMessagingCard'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { fetchAddressesAndValidate } from 'src/identity/actions'
import { AddressValidationType, SecureSendDetails } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { declinePaymentRequest } from 'src/paymentRequest/actions'
import { Recipient } from 'src/recipients/recipient'
import { RootState } from 'src/redux/reducers'
import { TransactionDataInput } from 'src/send/SendConfirmationLegacy'
import { Currency } from 'src/utils/currencies'
import Logger from 'src/utils/Logger'

interface Props {
  id: string
  requester: Recipient
  amount: string
  comment?: string
}

export default function IncomingPaymentRequestListItem({ id, amount, comment, requester }: Props) {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  const [payButtonPressed, setPayButtonPressed] = useState(false)
  const [addressesFetched, setAddressesFetched] = useState(false)

  const e164PhoneNumber = requester.e164PhoneNumber
  const requesterAddress = requester.address

  const secureSendDetails: SecureSendDetails | undefined = useSelector(
    (state: RootState) => state.identity.secureSendPhoneNumberMapping[e164PhoneNumber || '']
  )

  const onPayButtonPressed = () => {
    setPayButtonPressed(true)
    if (e164PhoneNumber) {
      // Need to check latest mapping to prevent user from accepting fraudulent requests
      dispatch(fetchAddressesAndValidate(e164PhoneNumber, requesterAddress))
    } else {
      navigateToNextScreen()
    }

    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.incoming_tx_request,
      selectedAction: NotificationBannerCTATypes.pay,
    })
  }

  const onDeclineButtonPressed = () => {
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.incoming_tx_request,
      selectedAction: NotificationBannerCTATypes.decline,
    })
    dispatch(declinePaymentRequest(id))
    Logger.showMessage(t('requestDeclined'))
  }

  const navigateToNextScreen = () => {
    const transactionData: TransactionDataInput = {
      reason: comment,
      recipient: requester,
      amount: new BigNumber(amount),
      currency: Currency.Dollar,
      type: TokenTransactionType.PayRequest,
      firebasePendingRequestUid: id,
    }

    const addressValidationType =
      secureSendDetails?.addressValidationType || AddressValidationType.NONE

    const origin = SendOrigin.AppRequestFlow
    if (addressValidationType === AddressValidationType.NONE) {
      navigate(Screens.SendConfirmationLegacy, { transactionData, origin })
    } else {
      navigate(Screens.ValidateRecipientIntro, {
        transactionData,
        addressValidationType,
        requesterAddress,
        origin,
      })
    }
  }

  useFocusEffect(
    useCallback(
      () => () => {
        // This is run when focus is lost.
        setPayButtonPressed(false)
        setAddressesFetched(false)
      },
      []
    )
  )

  useEffect(() => {
    // Need this to make sure it's only triggered on click
    if (!payButtonPressed) {
      return
    }

    const isFetchingAddresses = secureSendDetails?.isFetchingAddresses

    if (isFetchingAddresses) {
      setAddressesFetched(true)
    }

    if (addressesFetched && isFetchingAddresses === false) {
      setPayButtonPressed(false)
      if (secureSendDetails?.lastFetchSuccessful) {
        navigateToNextScreen()
      }
    }
  }, [payButtonPressed, secureSendDetails])

  return (
    <View style={styles.container}>
      <RequestMessagingCard
        testID={`IncomingPaymentRequestNotification/${id}`}
        title={t('incomingPaymentRequestNotificationTitle', { name: requester.name })}
        details={comment}
        amount={
          <CurrencyDisplay
            amount={{
              value: amount,
              currencyCode: Currency.Dollar,
            }}
          />
        }
        icon={<ContactCircle recipient={requester} />}
        callToActions={[
          {
            text: payButtonPressed ? (
              <ActivityIndicator testID={'loading/paymentRequest'} />
            ) : (
              t('send')
            ),
            onPress: onPayButtonPressed,
          },
          {
            text: t('decline'),
            onPress: onDeclineButtonPressed,
          },
        ]}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
})
