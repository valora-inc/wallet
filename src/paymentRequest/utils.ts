import { decryptComment, encryptComment } from '@celo/cryptographic-utils'
import { isE164NumberStrict } from '@celo/phone-utils'
import { hexToBuffer } from '@celo/utils/lib/address'
import BigNumber from 'bignumber.js'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import { Recipient } from 'src/recipients/recipient'
import { TransactionDataInput } from 'src/send/SendAmount'
import { TokenBalance } from 'src/tokens/slice'
import Logger from 'src/utils/Logger'
import { Currency } from 'src/utils/currencies'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Convert a payment request into a transaction data object if user has enough balance
export const transactionDataFromPaymentRequest = ({
  paymentRequest,
  stableTokens,
  requester,
}: {
  paymentRequest: PaymentRequest
  stableTokens: TokenBalance[]
  requester: Recipient
}): TransactionDataInput => {
  const cUsdTokenInfo = stableTokens.find((token) => token?.symbol === Currency.Dollar)
  const cEurTokenInfo = stableTokens.find((token) => token?.symbol === Currency.Euro)
  if (!cUsdTokenInfo?.address || !cEurTokenInfo?.address) {
    // Should never happen in production
    throw new Error('No token address found for cUSD or cEUR')
  }
  // If the user has enough cUSD balance, pay with cUSD
  // Else, try with cEUR
  // Else, throw error
  const usdRequested = new BigNumber(paymentRequest.amount)

  if (
    cUsdTokenInfo.usdPrice &&
    usdRequested.isLessThanOrEqualTo(cUsdTokenInfo.balance.multipliedBy(cUsdTokenInfo.usdPrice))
  ) {
    return {
      comment: paymentRequest.comment,
      recipient: requester,
      inputAmount: new BigNumber(paymentRequest.amount),
      tokenAmount: new BigNumber(paymentRequest.amount),
      amountIsInLocalCurrency: false,
      tokenAddress: cUsdTokenInfo.address,
      paymentRequestId: paymentRequest.uid || '',
    }
  } else if (
    cEurTokenInfo.usdPrice &&
    usdRequested.isLessThanOrEqualTo(cEurTokenInfo.balance.multipliedBy(cEurTokenInfo.usdPrice))
  ) {
    return {
      comment: paymentRequest.comment,
      recipient: requester,
      inputAmount: new BigNumber(paymentRequest.amount).dividedBy(cEurTokenInfo.usdPrice),
      tokenAmount: new BigNumber(paymentRequest.amount).dividedBy(cEurTokenInfo.usdPrice),
      amountIsInLocalCurrency: false,
      tokenAddress: cEurTokenInfo.address,
      paymentRequestId: paymentRequest.uid || '',
    }
  }
  throw new Error('Insufficient stablecoin balance')
}

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164NumberStrict(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable') ?? undefined
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
