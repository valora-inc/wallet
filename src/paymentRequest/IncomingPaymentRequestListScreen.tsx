import React from 'react'
import { WithTranslation } from 'react-i18next'
import { View } from 'react-native'
import { connect } from 'react-redux'
import i18n, { withTranslation } from 'src/i18n'
import { HeaderTitleWithBalance } from 'src/navigator/Headers'
import { NotificationList } from 'src/notifications/NotificationList'
import IncomingPaymentRequestListItem from 'src/paymentRequest/IncomingPaymentRequestListItem'
import { getIncomingPaymentRequests } from 'src/paymentRequest/selectors'
import { PaymentRequest } from 'src/paymentRequest/types'
import { getRecipientFromAddress, RecipientInfo } from 'src/recipients/recipient'
import { recipientInfoSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'
import { Currency } from 'src/utils/currencies'

interface StateProps {
  paymentRequests: PaymentRequest[]
  recipientInfo: RecipientInfo
}

const mapStateToProps = (state: RootState): StateProps => ({
  paymentRequests: getIncomingPaymentRequests(state),
  recipientInfo: recipientInfoSelector(state),
})

type Props = WithTranslation & StateProps

export const listItemRenderer = (props: { recipientInfo: RecipientInfo }) => (
  request: PaymentRequest,
  key: number | undefined = undefined
) => (
  <View key={key}>
    <IncomingPaymentRequestListItem
      id={request.uid || ''}
      amount={request.amount}
      requester={getRecipientFromAddress(request.requesterAddress, props.recipientInfo)}
      comment={request.comment}
    />
  </View>
)

class IncomingPaymentRequestListScreen extends React.Component<Props> {
  static navigationOptions = () => ({
    headerTitle: (
      <HeaderTitleWithBalance title={i18n.t('incomingPaymentRequests')} token={Currency.Dollar} />
    ),
  })

  render = () => {
    return (
      <NotificationList
        items={this.props.paymentRequests}
        listItemRenderer={listItemRenderer(this.props)}
      />
    )
  }
}

export default connect<StateProps, {}, {}, RootState>(
  mapStateToProps,
  {}
)(withTranslation<Props>()(IncomingPaymentRequestListScreen))
