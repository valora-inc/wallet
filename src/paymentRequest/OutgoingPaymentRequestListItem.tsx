import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import ContactCircle from 'src/components/ContactCircle'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import RequestMessagingCard from 'src/components/RequestMessagingCard'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { withTranslation } from 'src/i18n'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import { getDisplayName, Recipient } from 'src/recipients/recipient'
import { Currency } from 'src/utils/currencies'
import Logger from 'src/utils/Logger'

interface OwnProps {
  id: string
  requestee: Recipient
  amount: string
  comment?: string
  cancelPaymentRequest: typeof cancelPaymentRequest
  updatePaymentRequestNotified: typeof updatePaymentRequestNotified
  index?: number
}

type Props = OwnProps & WithTranslation

export class OutgoingPaymentRequestListItem extends React.Component<Props> {
  onRemind = () => {
    const { id, t, index } = this.props
    this.props.updatePaymentRequestNotified(id, false)
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.remind,
      notificationPositionInList: index,
    })
    Logger.showMessage(t('reminderSent'))
  }

  onCancel = () => {
    const { id, index } = this.props
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.decline,
      notificationPositionInList: index,
    })
    this.props.cancelPaymentRequest(id)
  }

  getCTA = () => {
    return [
      {
        text: this.props.t('remind'),
        onPress: this.onRemind,
      },
      {
        text: this.props.t('cancel'),
        onPress: this.onCancel,
      },
    ]
  }

  render() {
    const { requestee, id, comment, t } = this.props
    const amount = {
      value: this.props.amount,
      currencyCode: Currency.Dollar,
    }

    return (
      <RequestMessagingCard
        testID={`OutgoingPaymentRequestNotification/${id}`}
        title={t('outgoingPaymentRequestNotificationTitle', {
          name: getDisplayName(requestee, t),
        })}
        amount={<CurrencyDisplay amount={amount} />}
        details={comment}
        icon={<ContactCircle recipient={requestee} />}
        callToActions={this.getCTA()}
      />
    )
  }
}

export default withTranslation<Props>()(OutgoingPaymentRequestListItem)
