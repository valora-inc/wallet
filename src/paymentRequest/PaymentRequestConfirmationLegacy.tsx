import { firebase } from '@react-native-firebase/database'
import { StackScreenProps } from '@react-navigation/stack'
import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch } from 'react-redux'
import { e164NumberSelector } from 'src/account/selectors'
import { RequestEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import BackButton from 'src/components/BackButton'
import CommentTextInput from 'src/components/CommentTextInput'
import ContactCircle from 'src/components/ContactCircle'
import CurrencyDisplay, { DisplayType } from 'src/components/CurrencyDisplay'
import ReviewFrame from 'src/components/ReviewFrame'
import TotalLineItem from 'src/components/TotalLineItem'
import {
  e164NumberToAddressSelector,
  secureSendPhoneNumberMappingSelector,
} from 'src/identity/selectors'
import { emptyHeader } from 'src/navigator/Headers'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import { writePaymentRequest } from 'src/paymentRequest/actions'
import { PaymentRequestStatus } from 'src/paymentRequest/types'
import { getDisplayName } from 'src/recipients/recipient'
import useSelector from 'src/redux/useSelector'
import { getConfirmationInput } from 'src/send/utils'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import Logger from 'src/utils/Logger'
import { walletAddressSelector } from 'src/web3/selectors'

// @ts-ignore
const TAG = 'paymentRequest/confirmation'

type Props = StackScreenProps<StackParamList, Screens.PaymentRequestConfirmationLegacy>

export const paymentConfirmationScreenNavOptions = () => ({
  ...emptyHeader,
  headerLeft: () => <BackButton eventName={RequestEvents.request_confirm_back} />,
})

function PaymentRequestConfirmation({ route }: Props) {
  const [comment, setComment] = useState('')
  const { transactionData, addressJustValidated } = route.params
  const e164NumberToAddress = useSelector(e164NumberToAddressSelector)
  const secureSendPhoneNumberMapping = useSelector(secureSendPhoneNumberMappingSelector)
  const walletAddress = useSelector(walletAddressSelector)
  const requesterE164Number = useSelector(e164NumberSelector)

  const confirmationInput = getConfirmationInput(
    transactionData,
    e164NumberToAddress,
    secureSendPhoneNumberMapping
  )

  const { t } = useTranslation()

  useEffect(() => {
    if (addressJustValidated) {
      Logger.showMessage(t('addressConfirmed'))
    }
  }, [])

  const dispatch = useDispatch()

  const onBlur = () => {
    const trimmedComment = comment.trim()
    setComment(trimmedComment)
  }

  const onConfirm = async () => {
    const { amount, recipient, recipientAddress: requesteeAddress } = confirmationInput

    if (!recipient) {
      throw new Error("Can't request without valid recipient")
    }

    if (!walletAddress) {
      throw new Error("Can't request without a valid account")
    }

    if (!requesteeAddress) {
      throw new Error('Error passing through the requestee address')
    }

    const paymentInfo = {
      amount: amount.toString(),
      comment: comment || undefined,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      requesterAddress: walletAddress,
      requesterE164Number: requesterE164Number ?? undefined,
      requesteeAddress: requesteeAddress.toLowerCase(),
      status: PaymentRequestStatus.REQUESTED,
      notified: false,
    }

    ValoraAnalytics.track(RequestEvents.request_confirm_request, { requesteeAddress })
    dispatch(writePaymentRequest(paymentInfo))
    Logger.showMessage(t('requestSent'))
  }

  const amount = {
    value: confirmationInput.amount,
    currencyCode: confirmationInput.currency,
  }

  const renderFooter = () => {
    return (
      <View style={styles.feeContainer}>
        <TotalLineItem amount={amount} showExchangeRate={false} />
      </View>
    )
  }

  return (
    <SafeAreaView style={styles.container} edges={['bottom']}>
      <DisconnectBanner />
      <ReviewFrame
        FooterComponent={renderFooter}
        confirmButton={{
          action: onConfirm,
          text: t('request'),
          disabled: false,
        }}
      >
        <View style={styles.transferContainer}>
          <View style={styles.headerContainer}>
            <ContactCircle recipient={confirmationInput.recipient} />
            <View style={styles.recipientInfoContainer}>
              <Text style={styles.headerText}>{t('requesting')}</Text>
              <Text style={styles.displayName}>
                {getDisplayName(confirmationInput.recipient, t)}
              </Text>
            </View>
          </View>
          <CurrencyDisplay type={DisplayType.Default} style={styles.amount} amount={amount} />
          <CommentTextInput
            testID={'request'}
            onCommentChange={setComment}
            comment={comment}
            onBlur={onBlur}
          />
        </View>
      </ReviewFrame>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 8,
  },
  feeContainer: {
    padding: 16,
    paddingBottom: 8,
  },
  transferContainer: {
    alignItems: 'flex-start',
    paddingBottom: 24,
  },
  headerContainer: {
    flexDirection: 'row',
    alignItems: 'flex-start',
  },
  recipientInfoContainer: {
    paddingLeft: 8,
  },
  headerText: {
    ...fontStyles.regular,
    color: colors.gray4,
  },
  displayName: {
    ...fontStyles.regular500,
  },
  amount: {
    paddingVertical: 8,
    ...fontStyles.largeNumber,
  },
})

export default PaymentRequestConfirmation
