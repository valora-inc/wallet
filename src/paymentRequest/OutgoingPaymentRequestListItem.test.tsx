import { fireEvent, render } from '@testing-library/react-native'
import { noop } from 'lodash'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { NotificationBannerCTATypes, NotificationType } from 'src/home/types'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import OutgoingPaymentRequestListItem from 'src/paymentRequest/OutgoingPaymentRequestListItem'
import { RecipientType } from 'src/recipients/recipient'
import { createMockStore } from 'test/utils'

jest.mock('src/analytics/ValoraAnalytics')

const store = createMockStore()

const commonProps = {
  id: '1',
  amount: '24',
  comment: 'Hey thanks for the loan, Ill pay you back ASAP. LOVE YOU',
  requestee: {
    e164PhoneNumber: '5126608970',
    displayId: '5126608970',
    address: '0x91623f625e23ac1400',
    name: '5126608970',
    contact: undefined,
    recipientType: RecipientType.Address,
  },
  cancelPaymentRequest: noop as typeof cancelPaymentRequest,
  updatePaymentRequestNotified: noop as typeof updatePaymentRequestNotified,
  index: 4,
}

describe('OutgoingPaymentRequestListItem', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <OutgoingPaymentRequestListItem {...commonProps} />
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  it('emits correct analytics event when CTA button is pressed', () => {
    const { getByText } = render(
      <Provider store={store}>
        <OutgoingPaymentRequestListItem {...commonProps} />
      </Provider>
    )

    fireEvent.press(getByText('remind'))

    expect(ValoraAnalytics.track).toHaveBeenCalledTimes(1)
    expect(ValoraAnalytics.track).toHaveBeenLastCalledWith(HomeEvents.notification_select, {
      notificationType: NotificationType.outgoing_tx_request,
      notificationId: `${NotificationType.outgoing_tx_request}/1`,
      selectedAction: NotificationBannerCTATypes.remind,
      notificationPositionInList: 4,
    })
  })

  it('emits correct analytics event when notification is dismissed', () => {
    const { getByText } = render(
      <Provider store={store}>
        <OutgoingPaymentRequestListItem {...commonProps} />
      </Provider>
    )

    fireEvent.press(getByText('cancel'))

    expect(ValoraAnalytics.track).toHaveBeenCalledTimes(1)
    expect(ValoraAnalytics.track).toHaveBeenLastCalledWith(HomeEvents.notification_select, {
      notificationType: NotificationType.outgoing_tx_request,
      notificationId: `${NotificationType.outgoing_tx_request}/1`,
      selectedAction: NotificationBannerCTATypes.decline,
      notificationPositionInList: 4,
    })
  })
})
