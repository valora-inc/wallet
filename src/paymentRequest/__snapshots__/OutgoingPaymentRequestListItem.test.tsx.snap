// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutgoingPaymentRequestListItem renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "padding": 16,
      },
      {
        "borderRadius": 8,
      },
      undefined,
      [
        {
          "borderColor": "#E6E6E6",
          "borderWidth": 1,
          "flex": 1,
          "minHeight": 144,
        },
        {},
      ],
    ]
  }
  testID="OutgoingPaymentRequestNotification/1"
>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "marginRight": 16,
        }
      }
    >
      <Text
        numberOfLines={2}
        style={
          {
            "color": "#666666",
            "fontFamily": "Inter-Regular",
            "fontSize": 13,
            "lineHeight": 16,
            "marginBottom": 4,
          }
        }
        testID="OutgoingPaymentRequestNotification/1/Title"
      >
        outgoingPaymentRequestNotificationTitle, {"name":"5126608970"}
      </Text>
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#2E3338",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 24,
            "lineHeight": 28,
            "marginBottom": 2,
          }
        }
        testID="OutgoingPaymentRequestNotification/1/Amount"
      >
        <Text
          style={
            [
              undefined,
              {
                "color": undefined,
              },
            ]
          }
          testID="undefined/value"
        >
          â‚±
          31.92
        </Text>
      </Text>
      <Text
        style={
          {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
        testID="OutgoingPaymentRequestNotification/1/Details"
      >
        Hey thanks for the loan, Ill pay you back ASAP. LOVE YOU
      </Text>
    </View>
    <View
      style={{}}
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "justifyContent": "center",
              },
              {
                "backgroundColor": "hsl(166, 53%, 93%)",
                "borderRadius": 20,
                "height": 40,
                "width": 40,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 16,
                },
                {
                  "color": "hsl(166, 67%, 24%)",
                  "fontSize": 20,
                },
              ]
            }
          >
            5
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
      }
    }
    testID="OutgoingPaymentRequestNotification/1/CallToActions"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        {
          "attribute": "selectableItemBackgroundBorderless",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="OutgoingPaymentRequestNotification/1/CallToActions/remind/Button"
    >
      <Text
        style={
          [
            {
              "color": "#1AB775",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 16,
              "lineHeight": 22,
            },
            {
              "fontSize": 14,
              "lineHeight": 16,
              "marginRight": 24,
              "minHeight": 16,
              "minWidth": 48,
            },
          ]
        }
      >
        remind
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        {
          "attribute": "selectableItemBackgroundBorderless",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="OutgoingPaymentRequestNotification/1/CallToActions/cancel/Button"
    >
      <Text
        style={
          [
            {
              "color": "#1AB775",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 16,
              "lineHeight": 22,
            },
            {
              "fontSize": 14,
              "lineHeight": 16,
              "marginRight": 24,
              "minHeight": 16,
              "minWidth": 48,
            },
          ]
        }
      >
        cancel
      </Text>
    </View>
  </View>
</View>
`;
